File: delayedqueue-jvm/src/main/kotlin/org/funfix/delayedqueue/jvm/DelayedQueueTimeConfig.kt
package org.funfix.delayedqueue.jvm

import java.time.Duration

/**
 * Time configuration for delayed queue operations.
 *
 * @property acquireTimeout maximum time to wait when acquiring/locking a message for processing
 * @property pollPeriod interval between poll attempts when no messages are available
 */
@JvmRecord
public data class DelayedQueueTimeConfig(val acquireTimeout: Duration, val pollPeriod: Duration) {
    public companion object {
        /** Default configuration for [DelayedQueueInMemory]. */
        @JvmField
        public val DEFAULT_IN_MEMORY: DelayedQueueTimeConfig =
            DelayedQueueTimeConfig(
                acquireTimeout = Duration.ofMinutes(5),
                pollPeriod = Duration.ofMillis(500),
            )

        /**
         * Default configuration for JDBC-based implementations, with longer acquire timeouts and
         * poll periods to reduce database load in production environments.
         */
        @JvmField
        public val DEFAULT_JDBC: DelayedQueueTimeConfig =
            DelayedQueueTimeConfig(
                acquireTimeout = Duration.ofMinutes(5),
                pollPeriod = Duration.ofSeconds(3),
            )

        /**
         * Default configuration for testing, with shorter timeouts and poll periods to speed up
         * tests.
         */
        @JvmField
        public val DEFAULT_TESTING: DelayedQueueTimeConfig =
            DelayedQueueTimeConfig(
                acquireTimeout = Duration.ofSeconds(30),
                pollPeriod = Duration.ofMillis(100),
            )

        /** Creates a time configuration with the specified values. */
        @JvmStatic
        public fun create(acquireTimeout: Duration, pollPeriod: Duration): DelayedQueueTimeConfig =
            DelayedQueueTimeConfig(acquireTimeout, pollPeriod)
    }
}

File: delayedqueue-scala/src/main/scala/org/funfix/delayedqueue/scala/DelayedQueueTimeConfig.scala
/*
 * Copyright 2026 Alexandru Nedelcu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.funfix.delayedqueue.scala

import org.funfix.delayedqueue.jvm
import scala.concurrent.duration.FiniteDuration

/** Time configuration for delayed queue operations.
  *
  * @param acquireTimeout
  *   maximum time to wait when acquiring/locking a message for processing
  * @param pollPeriod
  *   interval between poll attempts when no messages are available
  */
final case class DelayedQueueTimeConfig(
  acquireTimeout: FiniteDuration,
  pollPeriod: FiniteDuration
) {

  /** Converts this Scala DelayedQueueTimeConfig to a JVM
    * DelayedQueueTimeConfig.
    */
  def asJava: jvm.DelayedQueueTimeConfig =
    new jvm.DelayedQueueTimeConfig(
      java.time.Duration.ofMillis(acquireTimeout.toMillis),
      java.time.Duration.ofMillis(pollPeriod.toMillis)
    )
}

object DelayedQueueTimeConfig {

  /** Default configuration for DelayedQueueInMemory. */
  val DEFAULT_IN_MEMORY: DelayedQueueTimeConfig =
    fromJava(jvm.DelayedQueueTimeConfig.DEFAULT_IN_MEMORY)

  /** Default configuration for JDBC-based implementations, with longer acquire
    * timeouts and poll periods to reduce database load in production
    * environments.
    */
  val DEFAULT_JDBC: DelayedQueueTimeConfig =
    fromJava(jvm.DelayedQueueTimeConfig.DEFAULT_JDBC)

  /** Default configuration for testing, with shorter timeouts and poll periods
    * to speed up tests.
    */
  val DEFAULT_TESTING: DelayedQueueTimeConfig =
    fromJava(jvm.DelayedQueueTimeConfig.DEFAULT_TESTING)

  /** Converts a Java `DelayedQueueTimeConfig` to a Scala
    * [[DelayedQueueTimeConfig]].
    */
  def fromJava(javaConfig: jvm.DelayedQueueTimeConfig): DelayedQueueTimeConfig =
    DelayedQueueTimeConfig(
      acquireTimeout =
        FiniteDuration(
          javaConfig.acquireTimeout.toMillis,
          scala.concurrent.duration.MILLISECONDS
        ),
      pollPeriod =
        FiniteDuration(javaConfig.pollPeriod.toMillis, scala.concurrent.duration.MILLISECONDS)
    )
}
